File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        visibleNodes = {}
        
        def dfs(node, depth):
            if node is None:
                return
            
            if depth not in visibleNodes:
                visibleNodes[depth] = node
            
            dfs(node.right, depth + 1)
            dfs(node.left, depth + 1)
            
        dfs(root, 0)
        
        result = []
        depth = 0
        while depth in visibleNodes:
            result.append(visibleNodes[depth].val)
            depth += 1
            
        return result

File: 199-binary-tree-right-side-view/199-binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> view = new ArrayList<>();
        if (root == null) return view;
        
        Queue<TreeNode> q = new LinkedList<>();
        q.offer(root);
        
        int level = 1;
        while (!q.isEmpty()) {
            int qSize = q.size();
            for (int i = 0; i < qSize; i++) {
                TreeNode node = q.poll();
                if (view.size() < level) {
                    view.add(node.val);
                }
                
                if (node.right != null) q.offer(node.right);
                if (node.left != null) q.offer(node.left);
            }
            level++;
        }
        return view;
    }
}